website/content/posts/gating-image-promotion-on-gitlab.md

```diff
@@ -0,0 +1,382 @@
+```
+
+We also modified the `tag-latest` job to depend on the `secure` job instead of `docker-build`. This
+ensures the image is signed before it is tagged.
+
+Notice how no signing keys are required to sign the image. This is because we are leveraging
+Sigstore's [identity-based] signatures, also known as "keyless".
+
+With the modifications above the GitLab pipeline will produce an image that is signed. Users can
+then verify it accordingly.
+
+## Add Provenance
+
+Taking this a step further, let's also create a [SLSA Provenance] that provides some information
+about how the image was created. Then, let's associate this information with our image as a signed
+attestation.
+
+To generate the SLSA Provenance, we are going to use a simple `bash` script that resides at
+`scripts/generate.sh` in our git repository.
+
+```bash
+#!/usr/bin/env bash
+set -euo pipefail
+
+# List of available GitLab CI variables: https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
+
+cat <<EOF
+{
+  "buildDefinition": {
+   "buildType": "https://gitlab.com/lucarval/sign-attest-poc",
+   "resolvedDependencies": [
+    {
+     "uri": "git+${CI_PROJECT_URL}",
+     "digest": {
+      "sha1": "${CI_COMMIT_SHA}"
+     }
+    }
+   ]
+  },
+  "runDetails": {
+   "builder": {
+    "id": "${CI_RUNNER_ID}",
+    "version": {
+     "gitlab-runner": "${CI_RUNNER_REVISION}"
+    }
+   },
+   "metadata": {
+    "invocationID": "${CI_PIPELINE_ID}",
+    "startedOn": "${CI_PIPELINE_CREATED_AT}",
+    "finishedOn": "${CI_PIPELINE_CREATED_AT}"
+   }
+  }
+}
+EOF
+```
+
+Next, we use this script to generate the SLSA Provenance predicate, and change the previously added
+`secure` job to associate this information with our image.
+
+```yaml
+# [...]
+
+generate-provenance:
+  needs: [docker-build]
+  stage: process
+  image: registry.access.redhat.com/ubi9:latest
+  script:
+    - ./scripts/generate.sh > predicate.json
+  artifacts:
+    paths:
+      - predicate.json
+
+secure:
+  needs:  # <- added
+    - job: generate-provenance
+      artifacts: true
+  stage: process
+  # [...]
+  script:
+    - cosign sign ${DOCKER_IMAGE_NAME}
+    # added line below
+    - cosign attest --predicate predicate.json --type https://slsa.dev/provenance/v1 ${DOCKER_IMAGE_NAME}
+
+# [...]
+```
+
+Now our image is signed and attested! The consumers of this image are ecstatic.
+
+## Gate Promotion
+
+There is a gotcha in our pipeline! What if there is a bug in our `generate.sh` script? What if the
+signature, for whatever reason, is not created as expected? Any consumer of our image that practices
+supply chain security best practices will be prevented from consuming the image. Those poor users...
+
+To mitigate these issues, we introduce a gating step that validates the image, like consumers would,
+*before* promoting the image. As part of this validation, let's define a minimal Enterprise Contract
+policy configuration file, called `policy.yaml`, that captures the validity requirements:
```

Comment:
should this be _validation_ requirements?

Reply:
It does sound a bit odd.. I reworked the sentence like this:

> Let's define a minimal Enterprise Contract policy configuration
file, called `policy.yaml`, that captures the validation requirements